### PR TITLE
add codecov test coverage report

### PR DESCRIPTION
**Pull request by MaHaWo**
- add upload of coverage report to codecov
- add badges in readme
- add python 3.9 and 3.12 (oldest and current newest) instead of only 3.11 to actions. 3.9 is the one rasperry-debian runs on, so we should test on it imho

### CODE REVIEW

**Comment by MaHaWo** [comment]
do you have any best practices as to how high code coverage should minimally be in order to be useful?  I currently set the failure-threshold to 80, although that's kind of arbitrary. Also, I activated branch coverage because it's imho more important than line coverage as such, at least when line coverage is not close to full?

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -51,4 +56,13 @@ jobs:
         pactl load-module module-remap-source master=virtmic.monitor source_name=virtmic source_properties=device.description=Virtual_Microphone
 
         # test stuff
-        python3 -m pytest tests
\ No newline at end of file
+        python3 -m pytest --cov=./src/iSparrowRecord --cov-report=xml
+
+    - name: Upload coverage reports to Codecov
+      uses: codecov/codecov-action@v4
```
Comment (iulusoy) [inline .github/workflows/main.yml:62]: could add here that report is only uploaded for latest Python version

**Review by iulusoy — APPROVED** [review]
There is no general rule how much test coverage you should have. It depends on the code - if there is a lot of gui code for example, that's hard to test. I think 80 is a good target.

File: .github/workflows/main.yml (hunk 1 above)
Comment (MaHaWo) [reply .github/workflows/main.yml:62]: add the python version to the name. that ok?

**Pull request merged**